lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb

```diff
@@ -269,10 +282,151 @@ def build_spn(options = {})
+      type: credential.keyblock.enctype.value,
+      value: credential.keyblock.data.value
+    )
+
+    etypes = Set.new([ticket.enc_part.etype])
+    etypes << Rex::Proto::Kerberos::Crypto::Encryption::RC4_HMAC
+
+    tgs_options = {
+      store_credential_cache: options.fetch(:store_credential_cache) { store_credential_cache },
+      credential_cache_username: options[:impersonate],
+      pa_data: build_pa_for_user(
+        {
+          username: options[:impersonate],
+          session_key: session_key,
+          realm: realm
+        }
+      )
+    }
+
+    request_service_ticket(
+      session_key,
+      ticket,
+      realm,
+      client_name,
+      etypes,
+      expiry_time,
+      now,
+      sname,
+      tgs_options
+    )
+  end
+
+  # Request a service ticket to another service on behalf of a user
+  #
+  # @param [Rex::Proto::Kerberos::CredentialCache::Krb5CcacheCredential] The ccache credential from the TGT
+  # @param [Hash] options
+  # @option options [Rex::Proto::Kerberos::Model::PrincipalName] :sname The target service principal name.
+  # @option options [Rex::Proto::Kerberos::Model::Ticket] :tgs_ticket The service ticket to the first service.
+  #   It must have the forwardable flag set. This ticket can be obtained with #s4u2self.
+  # @option options [Boolean] :store_credential_cache Override the @store_credential_cache attribute
+  # @option options [String] :impersonate The name of the user to request a ticket on behalf of
+  # @return [Array] The new TGS ticket and the decrypted TGS credentials as a MIT Cache Credential
+  def s4u2proxy(credential, options = {})
+    realm = self.realm.upcase
+    sname = options.fetch(:sname)
+    client_name = username
+
+    now = Time.now.utc
+    expiry_time = now + 1.day
+
+    ticket = Rex::Proto::Kerberos::Model::Ticket.decode(credential.ticket.value)
+    session_key = Rex::Proto::Kerberos::Model::EncryptionKey.new(
+      type: credential.keyblock.enctype.value,
+      value: credential.keyblock.data.value
+    )
+
+    pa_pac_options_flags = Rex::Proto::Kerberos::Model::KdcOptionFlags.from_flags(
+      [
+        Rex::Proto::Kerberos::Model::PreAuthPacOptionsFlags::RESOURCE_BASED_CONSTRAINED_DELEGATION
+      ]
+    )
+    pa_pac_options = Rex::Proto::Kerberos::Model::PreAuthPacOptions.new(
+      flags: pa_pac_options_flags
+    )
+    pa_data_entry = Rex::Proto::Kerberos::Model::PreAuthDataEntry.new(
+      type: Rex::Proto::Kerberos::Model::PreAuthType::PA_PAC_OPTIONS,
+      value: pa_pac_options.encode
+    )
+
+    etypes = Set.new(etypes)
+    etypes << Rex::Proto::Kerberos::Crypto::Encryption::RC4_HMAC
+    etypes << Rex::Proto::Kerberos::Crypto::Encryption::DES_CBC_MD5
+    etypes << Rex::Proto::Kerberos::Crypto::Encryption::DES3_CBC_SHA1
+
+    tgs_options = {
+      pa_data: pa_data_entry,
+      additional_flags: Rex::Proto::Kerberos::Model::KdcOptionFlags::CNAME_IN_ADDL_TKT,
+      additional_tickets: options[:tgs_ticket],
```

Comment:
Would this be an array?
```suggestion
      additional_tickets: [options[:tgs_ticket]],
```

Reply:
There is some code in #request_service_ticket that convert this value to an array if if it is not already:
```ruby
    if options[:additional_tickets].present?
      additional_tickets = [options[:additional_tickets]] unless options[:additional_tickets].is_a?(::Enumerable)
      tgs_body_options[:additional_tickets] = additional_tickets
    end
```
But I agree, it makes more sense to directly use an array. I will also update the `additional_flags`. Thanks for the heads up!